### PR TITLE
Enabling op model interface for constraints and L1 usage.

### PR DESCRIPTION
**Pull request by mbezuljTT**
This PR plumbs OpModelInterface to the underlying tt-metal op queries for validation and L1 memory consumption. 

`TTNNOpModelInterface.td` getOpConstraints takes input(s) and output `TTNNLayoutAttr` and returns a tuple of three values:

1. A boolean indicating if the op is legal for the given input/output layouts.
2. If the op is legal, a tuple of three values representing the op memory L1 usage estimate in bytes.
   - The first value is the CB L1 peak allocation in bytes.
   - The second value is the Tensor L1 peak allocation in bytes.
   - The third value is the Output L1 buffer allocation in bytes.
3. If the op is illegal, a string describing the failure.

`TTNNOpModelInterface.cpp` implements hooks to the _wrapper library_ 'TTNNOpModelLib' (where metal API is). Per each op, implementation takes
- tensor shapes (`llvm::ArrayRef<>`) from its operands,
- worker grid (used for virtual to physical cores conversion),
- op specific params (like softmax dimension), and 
- with layouts `TTNNLayoutsAttr`
and pass them to the _wrapper library_ `TTNNOpModelLib`.

`TTNNOpModelLib` converts mlir structures to metal structures, and calls into underlying 'tt-metal' op interface.

Underlying `tt-metal` op interface `::ttnn::graph::query_op_constraints(..)` consumes a target op (e.g. 'ttnn::relu') and it's arguments in the order of op implemented ::invoke function that we are targeting. 

Implemented `SingletonDeviceContext` to avoid constant opening/closing device. This class should ensure opened device is a mockup device when it's implemented on the tt-metal side (https://github.com/tenstorrent/tt-metal/issues/14000)

Added 3 types of unit tests:
- TestConversion - tests conversion of the MLIR to TTNN types
- TestOpModelLib - tests interface to metal API
- TestOpModelInterface - tests interface built in metal ops

Due to differences in tt-metal and LLVM project setups (compiler standard, exceptions) these are implemented as the place Google unit test. Unlike other unit tests that are also Google unit tests but wrapped into LLVM (and invoked using llvm-lit).

As these tests require TT hardware (until mockup device is implemented), changed Build tt-mlir  op_model flavour to use n300 runners. 

Additionally, wired op model interface in the ShardSolver; mnist_sharded.mlir compiles and runs. @odjuricicTT confirmed found solution is the one we expected. 

Internal doc describing more details can be found [here](https://tenstorrent-my.sharepoint.com/:w:/p/mbezulj/ETC6JOzVU9dAhQjgIAiwGt8BdjbNdXmMw-fZTo7As1BVXw?e=kM7a3c)

### CODE REVIEW

File: lib/OpModel/TTNN/TTNNOpModelLib.cpp [hunk 1]
```diff
@@ -152,29 +217,372 @@ getMemoryConfig(const mlir::tt::ttnn::TTNNLayoutAttr &layout) {
...
+    std::tuple<::llvm::ArrayRef<int64_t>, ::mlir::tt::ttnn::TTNNLayoutAttr,
+               ::llvm::ArrayRef<int64_t>, ::mlir::tt::ttnn::TTNNLayoutAttr,
+               ::llvm::ArrayRef<int64_t>, ::mlir::tt::ttnn::TTNNLayoutAttr>,
+    ::ttnn::graph::QueryResponse,
+    std::function<::ttnn::graph::QueryResponse(
+        const std::tuple<
+            ::llvm::ArrayRef<int64_t>, ::mlir::tt::ttnn::TTNNLayoutAttr,
+            ::llvm::ArrayRef<int64_t>, ::mlir::tt::ttnn::TTNNLayoutAttr,
+            ::llvm::ArrayRef<int64_t>, ::mlir::tt::ttnn::TTNNLayoutAttr> &)>>
+    g_constraints_cache_add_op(
```
Comment (nobradovictt) [inline lib/OpModel/TTNN/TTNNOpModelLib.cpp:395]: Why did you implement caching in the first iteration? It's unnecessary complication. Is there any resource governance mechanism for cache, it can just grow out of control?

File: lib/Dialect/TTNN/Analysis/ShardSolver.cpp [hunk 2]
```diff
@@ -508,13 +508,6 @@ bool ShardSolver::checkShardCompatible(
   // TEMP : Dummy mock implementation, will be replaced.
   //
 
-  if (OpModel backend = dyn_cast<OpModel>(consumerOp)) {
```
Comment (nobradovictt) [inline lib/Dialect/TTNN/Analysis/ShardSolver.cpp:511]: How should call roughly look?

Comment (nobradovictt) [reply lib/Dialect/TTNN/Analysis/ShardSolver.cpp:511]: Its ranked tensor type now or shape plus layout?

File: lib/OpModel/TTNN/TTNNOpModelLib.cpp (hunk 1 above)
Comment (nobradovictt) [reply lib/OpModel/TTNN/TTNNOpModelLib.cpp:395]: I also think it should not be global cache, it should be owned and managed by compiler component using it(Optimizer/Corresponding analysis).

Comment (mbezuljTT) [reply lib/OpModel/TTNN/TTNNOpModelLib.cpp:395]: I'm fine to remove it.

File: lib/Dialect/TTNN/Analysis/ShardSolver.cpp (hunk 2 above)
Comment (mbezuljTT) [reply lib/Dialect/TTNN/Analysis/ShardSolver.cpp:511]: it's ranked tensor type now. underlying API to TTNNModelLib is shape plus layout. 

Do you have preference? 

People on the meeting yesterday didn't have a strong opinion.

**Comment by mbezuljTT** [comment]
Q: if validation fails, we get an exception message from tt-metal, should we wire this back to the API caller? shall we make IsLegal returns a tuple<bool, optional<std::string>> ?  @nobradovictt  @odjuricicTT

File: test/unittests/OpModel/TestOpModelLib.cpp [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+  for (size_t i = 0; i < array.size(); ++i) {
+    if (i != 0)
+      os << ", ";
+    os << array[i];
+  }
+  os << "]";
+  return os;
+}
+
+TEST_F(OpModelBase, ReluInterleaved) {
```
Comment (odjuricicTT) [inline test/unittests/OpModel/TestOpModelLib.cpp:65]: Can we also have tests that call the high level interface using RankedTensorTypes? Or is this also blocked by llvm lit problem?

**Comment by odjuricicTT** [comment]
> Q: if validation fails, we get an exception message from tt-metal, should we wire this back to the API caller? shall we make IsLegal returns a tuple<bool, optionalstd::string> ?

I think it does make sense in the long run, we had something similar on Buda. Tho i don't know what is the scope of a change like this? We should definitely prioritize having something working e2e first.

File: test/unittests/OpModel/TestOpModelLib.cpp (hunk 3 above)
Comment (mbezuljTT) [reply test/unittests/OpModel/TestOpModelLib.cpp:65]: We can do that as well. And it would make sense to test interface on the op level as well. Though in that case we are doubling amount of test and maintenance / therefore I would prefer to use std::vector<std::tuple<shape, layout>> instead of RankedTensorType; or just remove these tests

**Comment by mbezuljTT** [comment]
> > Q: if validation fails, we get an exception message from tt-metal, should we wire this back to the API caller? shall we make IsLegal returns a tuple<bool, optionalstd::string> ?
> 
> I think it does make sense in the long run, we had something similar on Buda. Tho i don't know what is the scope of a change like this? We should definitely prioritize having something working e2e first.

It's a simple change. The error message is still a human readable string, unusable to the compiler, but maybe usable to the people running/debugging compiler. Moving to program-friendly error message would be much harder problem.

File: test/unittests/OpModel/TTNN/Lib/TestOpModelLib.cpp [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+                   mlir::tt::ttnn::TensorMemoryLayout::BlockSharded, {4, 4});
+  mlir::tt::ttnn::TTNNLayoutAttr inputLayout_dram =
+      CreateLayout(tensorShape, mlir::tt::ttnn::BufferType::DRAM,
+                   mlir::tt::ttnn::TensorMemoryLayout::Interleaved, {4, 4});
+
+  std::tuple<bool, std::optional<std::tuple<size_t, size_t, size_t>>,
+             std::optional<std::string>>
+      opConstraint;
+  bool legal = false;
+  legal = std::get<0>(MatmulOpInterface::getOpConstraints(
```
Comment (nobradovictt) [inline test/unittests/OpModel/TTNN/Lib/TestOpModelLib.cpp:321]: Any test validation of L1 usage values?

File: .github/workflows/build-and-test.yml [hunk 5]
```diff
@@ -80,7 +80,7 @@ jobs:
         -DTTMLIR_ENABLE_RUNTIME=ON \
         -DTTMLIR_ENABLE_RUNTIME_TESTS=ON \
         -DTTMLIR_ENABLE_STABLEHLO=ON \
-        -DTTMLIR_ENABLE_OP_MODEL=ON \
+        -DTTMLIR_ENABLE_OPMODEL=ON \
```
Comment (vmilosevic) [inline .github/workflows/build-and-test.yml:83]: Is this change intentional?
TTMLIR_ENABLE_OP_MODEL -> TTMLIR_ENABLE_OPMODEL

**Review by vmilosevic — APPROVED** [review]

File: .github/workflows/build-and-test.yml (hunk 5 above)
Comment (mbezuljTT) [reply .github/workflows/build-and-test.yml:83]: yes this is intentional, it was a bug. thanks for checking!

**Comment by odjuricicTT** [comment]
> The third value is the Output L1 buffer allocation in bytes.

@mbezuljTT What will the third value be if the op is DPS (the output tensor is pre-allocated and passed in as an arg)?

**Review by odjuricicTT — COMMENTED** [review]
Went trought half the changes, will continue after lunch.

**Comment by mbezuljTT** [comment]
> > The third value is the Output L1 buffer allocation in bytes.
> 
> @mbezuljTT What will the third value be if the op is DPS (the output tensor is pre-allocated and passed in as an arg)?


Op query functions implemented at TTNNOpModelLib are not DPS, therefore, it would be size of the output tensor anyway. however, peak usage might be wrong in this case as it might include output tensor size (which is probably not what you want). 

when Ops really become DPS, you would want to change how op is invoked in the TTNNOpModelLib.cpp to DPS as well. When you do that, third value would become zero, but you can use another graph capture around create_device_tensor for output allocation to get it's size.

File: .github/actions/build-tt-mlir-action/action.yaml [hunk 6]
```diff
@@ -0,0 +1,110 @@
+name: "Build tt-mlir"
```
[the remark this answers is not in the document]
Comment (mbezuljTT) [reply .github/actions/build-tt-mlir-action/action.yaml:1]: when op model interface is enabled, everything is compiled in a slightly different manner; sharded mnist will not be compiled in the same way, therefore check-ttmlir would have different results. I think it's worth testing it?

with that in mind, please let me know if you still prefer me to move OpModelInterface to a specific step in the new job.

[the remark this answers is not in the document]
Comment (odjuricicTT) [reply .github/actions/build-tt-mlir-action/action.yaml:1]: Ok, that makes sense. Let's leave as is then, i agree that we want to test that as well.

File: lib/Dialect/TTNN/Analysis/ShardSolver.cpp [hunk 7]
```diff
@@ -520,10 +520,74 @@ bool ShardSolver::checkShardCompatible(
...
+
+    // map consumer operands to DRAM interleave or provided producerLayout
+    // only one operand can be mapped to producerLayout, it's picked as first
+    // operand matching producerOp output shape.
+
+    uint32_t numOperands = consumerOp->getNumOperands();
+    // some ops have multiple operands; and some ops have output also an
+    // operand. TBD if there is a more robust way to get real number of inputs
+
+    // cast to DPSop?  //
```
Comment (odjuricicTT) [inline lib/Dialect/TTNN/Analysis/ShardSolver.cpp:536]: Please leave a "TODO(odjuricic): cast to DPSop?"

File: lib/Dialect/TTNN/Analysis/ShardSolver.cpp [hunk 8]
```diff
@@ -520,10 +520,74 @@ bool ShardSolver::checkShardCompatible(
...
+            TileType::get(consumerOp->getContext(), input.getElementType());
+
+        auto layout = TTNNLayoutAttr::get(
+            consumerOp->getContext(), input.getShape(), elementType,
+            BufferType::DRAM, workerGrid,
+            TensorMemoryLayoutAttr::get(consumerOp->getContext(),
+                                        TensorMemoryLayout::Interleaved));
+        inputLayouts.push_back(layout);
+      }
+    }
```
Comment (odjuricicTT) [inline lib/Dialect/TTNN/Analysis/ShardSolver.cpp:567]: As discussed offline, a bit of cleanup is needed here. I'll do this in a follow up PR.

File: lib/Dialect/TTNN/Analysis/ShardSolver.cpp [hunk 9]
```diff
@@ -520,10 +520,74 @@ bool ShardSolver::checkShardCompatible(
...
+    if (debug) {
+      llvm::errs() << "OpModel constraints valid. ";
+      llvm::errs() << producerOp->getName() << "->" << consumerOp->getName()
+                   << "\n";
+      producerLayout.dump();
+      consumerLayout.dump();
+    }
   }
 
   // May need to fetch other inputs for consumerOp(weights/join node).
```
Comment (odjuricicTT) [inline lib/Dialect/TTNN/Analysis/ShardSolver.cpp:593]: Could you remove the check on line 634, and just see if mnist still works?

**Review by odjuricicTT — APPROVED** [review]
Looks good! Thanks for pushing this all the way :)

I have a few more details to look over tomorrow.

File: test/unittests/OpModel/TTNN/Lib/TestOpModelLib.cpp [hunk 10]
```diff
@@ -0,0 +1,660 @@
...
+      tensorShape, inputLayout_l1_hs, -2, tensorShape, inputLayout_l1_hs,
+      workerGrid);
+  std::tie(legal, l1Usage, errorMsg) = opConstraint;
+  EXPECT_TRUE(legal);
+  EXPECT_TRUE(l1Usage.has_value());
+  EXPECT_FALSE(errorMsg.has_value());
+  std::tie(cb_size, peak_size, output_size) = l1Usage.value();
+  EXPECT_EQ(cb_size, 24576);
+  EXPECT_EQ(output_size, 32768);
+  EXPECT_EQ(peak_size, 32768);
```
Comment (odjuricicTT) [inline test/unittests/OpModel/TTNN/Lib/TestOpModelLib.cpp:291]: Is this the same test as on line 245?

File: test/unittests/OpModel/TTNN/Lib/TestOpModelLib.cpp [hunk 11]
```diff
@@ -0,0 +1,660 @@
...
+  EXPECT_FALSE(errorMsg.has_value());
+  std::tie(cb_size, peak_size, output_size) = l1Usage.value();
+  EXPECT_EQ(cb_size, 786432);
+  EXPECT_EQ(output_size, 151552);
+  EXPECT_EQ(peak_size, 151552);
+}
+
+TEST_F(OpModelBase, MatmulSharded) {
+  const llvm::SmallVector<int64_t> tensorShape = {1024, 1024};
+  const SmallVector<int64_t> gridShape = {4, 4};
```
Comment (odjuricicTT) [inline test/unittests/OpModel/TTNN/Lib/TestOpModelLib.cpp:602]: Is there a reason for not using 8x8 here?

File: test/unittests/OpModel/TTNN/Lib/TestOpModelLib.cpp [hunk 12]
```diff
@@ -0,0 +1,660 @@
...
+  const auto workerGrid = CreateWorkerGrid(
+      mlir::tt::ttnn::TensorMemoryLayout::BlockSharded, gridShape);
+  const mlir::tt::ttnn::TTNNLayoutAttr layout_l1_hs =
+      CreateLayout(tensorShape, mlir::tt::ttnn::BufferType::L1,
+                   mlir::tt::ttnn::TensorMemoryLayout::BlockSharded, gridShape);
```
Comment (odjuricicTT) [inline test/unittests/OpModel/TTNN/Lib/TestOpModelLib.cpp:607]: Name says _hs, but BlockSharded is passed.

File: test/unittests/OpModel/TTNN/Lib/TestOpModelLib.cpp [hunk 13]
```diff
@@ -0,0 +1,660 @@
...
+  opConstraint = (MatmulOpInterface::getOpConstraints(
+      tensorShape, layout_dram, tensorShape, layout_dram, tensorShape,
+      layout_l1_hs, false, false, workerGrid));
+  std::tie(legal, l1Usage, errorMsg) = opConstraint;
+  EXPECT_TRUE(legal);
+  EXPECT_TRUE(l1Usage.has_value());
+  EXPECT_FALSE(errorMsg.has_value());
+  std::tie(cb_size, peak_size, output_size) = l1Usage.value();
+  EXPECT_EQ(cb_size, 262144);
+  EXPECT_EQ(output_size, 524288);
```
Comment (odjuricicTT) [inline test/unittests/OpModel/TTNN/Lib/TestOpModelLib.cpp:656]: This seems wrong. It is 4 times the output of previous example. Am i missing something?

One possible explanation is that the output did not get sharded to a 4x4 grid for some reason, but the op chose a 1x4.
Can you imagine any possible way of reading the grid that the op chose (or other params) using graph capture? (not the scope of this PR ofc, just curious)

File: test/unittests/OpModel/TTNN/Lib/TestOpModelLib.cpp [hunk 14]
```diff
@@ -0,0 +1,660 @@
...
+  EXPECT_EQ(peak_size, 524288);
+}
```
Comment (odjuricicTT) [inline test/unittests/OpModel/TTNN/Lib/TestOpModelLib.cpp:658]: Could we test a few more cases for matmul?
* Hight sharded
* Block sharded for both inputs and output?
* Mix of block / hight sharding (this shouldn't work?)

File: test/unittests/OpModel/TTNN/Op/TestOpModelInterface.cpp [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+  std::optional<
+      std::tuple<bool, std::optional<std::tuple<size_t, size_t, size_t>>,
+                 std::optional<std::string>>>
+  getOpConstraints(Operation *op) {
+    std::vector<TTNNLayoutAttr> inputs;
+
+    // create input layouts
+    auto numOperand = op->getNumOperands();
+    // some ops have multiple operands
+    auto limit = (numOperand > 1) ? numOperand - 1 : numOperand;
```
Comment (odjuricicTT) [inline test/unittests/OpModel/TTNN/Op/TestOpModelInterface.cpp:29]: Leave a TODO here as well to check for DPS explicitly.

File: lib/Dialect/TTNN/Analysis/ShardSolver.cpp (hunk 9 above)
Comment (mbezuljTT) [reply lib/Dialect/TTNN/Analysis/ShardSolver.cpp:593]: removed, it works

File: test/unittests/OpModel/TTNN/Lib/TestOpModelLib.cpp [hunk 16]
```diff
@@ -0,0 +1,660 @@
...
+
+  opConstraint = ReluOpInterface::getOpConstraints(
+      tensorShape, inputLayout_dram, tensorShape, inputLayout_l1, workerGrid);
+  std::tie(legal, l1Usage, errorMsg) = opConstraint;
+  EXPECT_TRUE(legal);
+  EXPECT_TRUE(l1Usage.has_value());
+  EXPECT_FALSE(errorMsg.has_value());
+  std::tie(cb_size, peak_size, output_size) = l1Usage.value();
+  EXPECT_EQ(cb_size, 8192);
+  EXPECT_EQ(output_size, 4096);
```
[the remark this answers is not in the document]
Comment (mbezuljTT) [reply test/unittests/OpModel/TTNN/Lib/TestOpModelLib.cpp:67]: x2, double buffered

File: include/ttmlir/Dialect/TT/Utils/CoreRangeSet.h [hunk 17]
```diff
@@ -0,0 +1,75 @@
+// SPDX-FileCopyrightText: (c) 2024 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef TTMLIR_DIALECT_TT_UTILS_CORERANGESET_H
+#define TTMLIR_DIALECT_TT_UTILS_CORERANGESET_H
+
+#include "ttmlir/Dialect/TT/IR/TTOpsTypes.h"
+#include "ttmlir/Utils.h"
+#include <tuple>
```
Comment (svuckovicTT) [inline include/ttmlir/Dialect/TT/Utils/CoreRangeSet.h:10]: nit: add newline

File: include/ttmlir/Dialect/TTNN/IR/TTNNOpModelInterface.td [hunk 18]
```diff
@@ -19,32 +19,25 @@ def TTNN_OpModelInterface : OpInterface<"OpModel"> {
             }],
             /*retTy=*/"size_t",
             /*methodName=*/"getOpPerfCycles",
-            /*args=*/(ins "const std::vector<TTNNLayoutAttr>&":$input_layouts, "const TTNNLayoutAttr&":$output_layout),
+            /*args=*/(ins "const std::vector<TTNNLayoutAttr>&":$inputs, "const TTNNLayoutAttr&":$output),
```
Comment (svuckovicTT) [inline include/ttmlir/Dialect/TTNN/IR/TTNNOpModelInterface.td:22]: There are some ops that might return multiple outputs. I know that maxpool2d can return the output tensor + indices of each chosen pixel in sliding window, needed for backwards pass.

File: include/ttmlir/Dialect/TT/Utils/CoreRangeSet.h [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+
+namespace mlir::tt::utils {
+
+using locsize2d = std::tuple<std::array<int32_t, 2>,  //  {{locX, locY},
```
Comment (svuckovicTT) [inline include/ttmlir/Dialect/TT/Utils/CoreRangeSet.h:14]: nit: `locsize` -> `locSize`

File: test/unittests/OpModel/TTNN/Lib/TestOpModelLib.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (mbezuljTT) [reply test/unittests/OpModel/TTNN/Lib/TestOpModelLib.cpp:67]: N300 by default has 56 worker cores and 56 interleaved banks. It can have 64 when ethernet cores are used for dispatch.
Thus some cores will have two tiles, and as allocation always goes from bank 0, it's effectively the same as all cores having two tiles.

File: include/ttmlir/Dialect/TTNN/IR/TTNNOpModelInterface.td [hunk 20]
```diff
@@ -19,32 +19,25 @@ def TTNN_OpModelInterface : OpInterface<"OpModel"> {
...
             /*desc=*/[{
-                Returns the op memory L1 usage estimate in bytes. The return value is a tuple of 3 values:
-                - The first value is CB L1 peak allocation in bytes.
-                - The second value is Tensor L1 peak allocation in bytes.
-                - The third value is Output L1 buffer allocation in bytes.
+                Returns a tuple of three values:**
+                1. A boolean indicating if the op is legal for the given input/output layouts.
+                2. If the op is legal, a tuple of three values representing the op memory L1 usage estimate in bytes.
+                   - The first value is the CB L1 peak allocation in bytes.
+                   - The second value is the Tensor L1 peak allocation in bytes.
```
Comment (svuckovicTT) [inline include/ttmlir/Dialect/TTNN/IR/TTNNOpModelInterface.td:32]: What's the diff between CB and Tensor?

File: test/unittests/OpModel/TTNN/Lib/TestOpModelLib.cpp [hunk 21]
```diff
@@ -0,0 +1,660 @@
...
+  opConstraint = ReluOpInterface::getOpConstraints(
+      tensorShape, inputLayout_l1_hs, tensorShape, inputLayout_l1_hs,
+      workerGrid);
+  std::tie(legal, l1Usage, errorMsg) = opConstraint;
+  EXPECT_TRUE(legal);
+  EXPECT_TRUE(l1Usage.has_value());
+  EXPECT_FALSE(errorMsg.has_value());
+  std::tie(cb_size, peak_size, output_size) = l1Usage.value();
+  EXPECT_EQ(cb_size, 0);
+  EXPECT_EQ(output_size, tensorShape[0] * tensorShape[1] * 2 / 64);
```
[the remark this answers is not in the document]
Comment (mbezuljTT) [reply test/unittests/OpModel/TTNN/Lib/TestOpModelLib.cpp:125]: that's not entirely and always true; 

there are two folds to this observation:
- grid phy cores is really 7x8 on this N300, not 64. I need to add a check of provided workerGrid versus a real device.
- never the less, as interleaved allocation always starts with bank 0, it can take up to num_banks -1 more memory than equivalent sharded tensor; 

in this specific case this appeared as x2.

File: include/ttmlir/OpModel/TTNN/TTNNOpModel.h [hunk 22]
```diff
@@ -7,18 +7,71 @@
 
 #include "ttmlir/Dialect/TTNN/IR/TTNNOpsAttrs.h"
 
+#include <llvm/ADT/ArrayRef.h>
 #include <tuple>
```
Comment (svuckovicTT) [inline include/ttmlir/OpModel/TTNN/TTNNOpModel.h:11]: `<>` -> `""` for the llvm include, add newline between these.

https://docs.tenstorrent.com/tt-mlir/coding-guidelines.html#includes

File: test/unittests/OpModel/TTNN/Lib/TestOpModelLib.cpp (hunk 10 above)
Comment (mbezuljTT) [reply test/unittests/OpModel/TTNN/Lib/TestOpModelLib.cpp:291]: yes, I'll remove

File: include/ttmlir/Dialect/TTNN/Utils/VirtualToPhysicalGrid.h [hunk 23]
```diff
@@ -0,0 +1,76 @@
...
+/// - BlockSharded: Maps a block-sharded virtual grid (MxN) directly to a
+///   physical grid with the specified shape.
+///
+/// \param context The MLIR context.
+/// \param tensorMemoryLayout The tensor memory layout type.
+/// \param physicalGridShape The shape of the physical grid, defaults to {8, 8}.
+///
+/// \return An affine map that translates the virtual grid layout to the
+/// physical grid layout based on the specified tensor memory layout.
+auto SingleDeviceCreateVirtualToPhysicalLayoutMap(
```
Comment (svuckovicTT) [inline include/ttmlir/Dialect/TTNN/Utils/VirtualToPhysicalGrid.h:31]: Please go thru the whole PR and fix accordingly:

https://docs.tenstorrent.com/tt-mlir/coding-guidelines.html#using-auto-to-deduce-type

File: include/ttmlir/OpModel/TTNN/TTNNOpModel.h (hunk 22 above)
Comment (svuckovicTT) [reply include/ttmlir/OpModel/TTNN/TTNNOpModel.h:11]: There are a few other files with similar issues, please go through all of them.

File: lib/Dialect/TTNN/Analysis/ShardSolver.cpp [hunk 24]
```diff
@@ -520,10 +520,74 @@ bool ShardSolver::checkShardCompatible(
...
 
   if (OpModel backend = dyn_cast<OpModel>(consumerOp)) {
-    if (false ==
-        backend.isOpLegal(std::vector{producerLayout}, consumerLayout)) {
+
+    auto deviceAttr = mlir::tt::getCurrentScopeDevice(producerOp);
+    assert(deviceAttr);
+    auto workerGrid = deviceAttr.getWorkerGrid();
+
+    // map consumer operands to DRAM interleave or provided producerLayout
```
Comment (svuckovicTT) [inline lib/Dialect/TTNN/Analysis/ShardSolver.cpp:528]: https://docs.tenstorrent.com/tt-mlir/coding-guidelines.html#comments

File: test/unittests/OpModel/TTNN/Lib/TestOpModelLib.cpp (hunk 11 above)
Comment (mbezuljTT) [reply test/unittests/OpModel/TTNN/Lib/TestOpModelLib.cpp:602]: Matmul does check actual grid size and thus fails
 Number of rows in matmul compute grid exceeds maximum device compute grid size!


it works fine with 7x8

File: lib/OpModel/TTNN/Conversion.cpp [hunk 25]
```diff
@@ -0,0 +1,160 @@
...
+        CoreCoord(loc[0] + size[0] - 1, loc[1] + size[1] - 1)));
+  }
+  return ::tt::tt_metal::CoreRangeSet(coreRangeSet);
+}
+
+std::optional<::tt::tt_metal::ShardSpec>
+getShardSpec(const mlir::tt::ttnn::TTNNLayoutAttr &layout,
+             const mlir::tt::GridAttr &workerGrid) {
+  // tt_ShardOrientation is not part of ttnn::TTNNLayoutAttr;
+  // defaulting to ROW_MAJOR. TODO: figure out if we need to expose this
```
Comment (svuckovicTT) [inline lib/OpModel/TTNN/Conversion.cpp:86]: `ShardSpec` defines the orientation. There's an outstanding ticket with the goal of adding a `ShardSpecAttr` to `MemoryConfigAttr`.

https://github.com/tenstorrent/tt-mlir/issues/620

File: lib/OpModel/TTNN/Conversion.cpp [hunk 26]
```diff
@@ -0,0 +1,160 @@
...
+             ? std::make_optional(ShardSpec(getCoreRangeSet(layout, workerGrid),
+                                            getShardShape(layout),
+                                            ShardOrientation::ROW_MAJOR, false))
+             : std::nullopt;
+}
+
+::tt::tt_metal::BufferType getBufferType(const mlir::MemRefType &memref) {
+  auto memorySpace =
+      mlir::cast<mlir::tt::ttnn::BufferTypeAttr>(memref.getMemorySpace())
+          .getValue();
```
Comment (svuckovicTT) [inline lib/OpModel/TTNN/Conversion.cpp:97]: There is a `getBufferType()` in `lib/Dialect/TTNN/IR/TTNNOpsAttrs.cpp`. Consider exposing it in header and call that here.

File: test/lit.cfg.py [hunk 27]
```diff
@@ -98,3 +98,16 @@ def set_system_desc_features(system_desc):
...
+    llvm_config.with_environment("TT_METAL_HOME", os.environ["TT_METAL_HOME"])
+else:
+    print("Error: TT_METAL_HOME not set")
+
+if "ARCH_NAME" in os.environ:
+    print(f"ARCH_NAME={os.environ['ARCH_NAME']}")
+    llvm_config.with_environment("ARCH_NAME", os.environ["ARCH_NAME"])
+else:
+    print("ARCH_NAME not set. Defaulting to wormhole")
+    llvm_config.with_environment("ARCH_NAME", "wormhole_b0")
```
Comment (svuckovicTT) [inline test/lit.cfg.py:113]: I would argue that we should error out here as well. In Buda we used to have defaults like this and it only made it harder to debug what was happening when there was a problem with env setup.

File: test/lit.cfg.py [hunk 28]
```diff
@@ -98,3 +98,16 @@ def set_system_desc_features(system_desc):
     ],
     append_path=True,
 )
+
+if "TT_METAL_HOME" in os.environ:
```
Comment (svuckovicTT) [inline test/lit.cfg.py:102]: Would you mind adding `TT_MLIR_HOME` as well please?

File: test/ttmlir/Silicon/TTNN/optimizer/mnist_sharding.mlir [hunk 29]
```diff
@@ -1,11 +1,11 @@
-// RUN: ttmlir-opt --ttir-to-ttnn-backend-pipeline="system-desc-path=%system_desc_path% enable-optimizer=true memory-layout-analysis-enabled=true" %s > %t.mlir
-// RUN: FileCheck %s --input-file=%t.mlir
-// RUN: ttmlir-translate --ttnn-to-flatbuffer %t.mlir > %t.ttnn
+// RUN: ttmlir-opt --ttir-to-ttnn-backend-pipeline="system-desc-path=%system_desc_path% enable-optimizer=true memory-layout-analysis-enabled=true" -o output_file.mlir %s
```
Comment (svuckovicTT) [inline test/ttmlir/Silicon/TTNN/optimizer/mnist_sharding.mlir:1]: I'm curious, why did you change from piping to using `-o` arg?

**Review by svuckovicTT — APPROVED** [review]
Commented mostly on cosmetic stuff. Dialect side, changes look good.

NB: We don't have `ShardSpec` modelled in the TTNN dialect today, but it will soon be added (🤞), how much work do you think it'll be to adjust these changes to reflect that?

File: test/unittests/OpModel/TTNN/Lib/TestOpModelLib.cpp (hunk 13 above)
Comment (mbezuljTT) [reply test/unittests/OpModel/TTNN/Lib/TestOpModelLib.cpp:656]: Good catch; I need to investigate this one.

File: test/ttmlir/Silicon/TTNN/optimizer/mnist_sharding.mlir (hunk 29 above)
Comment (mbezuljTT) [reply test/ttmlir/Silicon/TTNN/optimizer/mnist_sharding.mlir:1]: op model interface prints a lot of things, as it's actually running metal code.

File: test/unittests/OpModel/TTNN/Lib/TestOpModelLib.cpp (hunk 13 above)
Comment (mbezuljTT) [reply test/unittests/OpModel/TTNN/Lib/TestOpModelLib.cpp:656]: this is a proper metal bug, for some reason output is sharded on less cores.

File: include/ttmlir/Dialect/TTNN/IR/TTNNOpModelInterface.td (hunk 18 above)
Comment (mbezuljTT) [reply include/ttmlir/Dialect/TTNN/IR/TTNNOpModelInterface.td:22]: good call. Added todo, note: @odjuricicTT

File: include/ttmlir/Dialect/TTNN/IR/TTNNOpModelInterface.td (hunk 20 above)
Comment (mbezuljTT) [reply include/ttmlir/Dialect/TTNN/IR/TTNNOpModelInterface.td:32]: circular buffers are used for communication between riscv cores within tensixes; and for shards. while tensors are higher level concept. they are allocated from different sides of the memory.

File: lib/OpModel/TTNN/Conversion.cpp (hunk 25 above)
Comment (mbezuljTT) [reply lib/OpModel/TTNN/Conversion.cpp:86]: thanks, added todo
